ecs-init/volumes/ecs_volume_plugin.go

```diff
@@ -201,32 +188,148 @@ func deleteMountPath(path string) error {
 func (a *AmazonECSVolumePlugin) Mount(r *volume.MountRequest) (*volume.MountResponse, error) {
-	a.lock.RLock()
-	defer a.lock.RUnlock()
+	seelog.Infof("Received mount request %+v", r)
+
+	// Validate the request
+	if len(r.Name) == 0 {
+		return nil, fmt.Errorf("no volume in the request")
+	}
+	if len(r.ID) == 0 {
+		return nil, fmt.Errorf("no mount ID in the request")
+	}
+
+	// Acquire write lock
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	// Find the volume
 	vol, ok := a.volumes[r.Name]
 	if !ok {
 		seelog.Errorf("Volume %s to mount is not found", r.Name)
 		return nil, fmt.Errorf("volume %s not found", r.Name)
 	}
+
+	// Find the volume driver
+	volDriver, err := a.getVolumeDriver(vol.Type)
+	if err != nil {
+		seelog.Errorf("Volume %s's driver type %s not supported: %v", r.Name, vol.Type, err)
+		return nil, fmt.Errorf("Volume %s's driver type %s not supported: %w", r.Name, vol.Type, err)
+	}
+	if volDriver == nil {
+		// This case shouldn't happen normally
+		return nil, fmt.Errorf("no volume driver found for type %s", vol.Type)
+	}
+
+	// Mount the volume on the host if there are no active mounts for the volume.
+	if len(vol.Mounts) == 0 {
+		seelog.Infof("Mounting volume %s as there are no existing mounts for it", r.Name)
+		createReq := &driver.CreateRequest{Name: r.Name, Path: vol.Path, Options: vol.Options}
+		if err := volDriver.Create(createReq); err != nil {
+			seelog.Errorf("Volume %s creation failure: %v", r.Name, err)
+			return nil, fmt.Errorf("failed to mount volume %s: %w", r.Name, err)
+		}
+		seelog.Infof("Volume %s mounted successfully", r.Name)
+	}
+
+	// Update state
+	seelog.Infof("Adding mount %s to volume %s", r.ID, r.Name)
+	vol.AddMount(r.ID)
+	if err := a.state.recordVolume(r.Name, vol); err != nil {
+		// State update failed, so roll back the changes made so far to make state consistent
+		seelog.Errorf("Failed to save volume %s, rolling back changes: %v", r.Name, err)
+		vol.RemoveMount(r.ID)
+		if len(vol.Mounts) == 0 {
+			seelog.Warnf("Rolling back mounting of volume %s", r.Name)
+			if err := volDriver.Remove(&driver.RemoveRequest{Name: r.Name}); err != nil {
+				seelog.Errorf("Volume %s removal failure: %v", r.Name, err)
+			}
+		}
+		a.state.recordVolume(r.Name, vol)
+		return nil, fmt.Errorf("mount failed due to an error while saving state: %w", err)
+	}
+
+	// All good
 	return &volume.MountResponse{Mountpoint: vol.Path}, nil
 }
 
 // Unmount implements Docker volume plugin's Unmount Method
 func (a *AmazonECSVolumePlugin) Unmount(r *volume.UnmountRequest) error {
-	a.lock.RLock()
-	defer a.lock.RUnlock()
-	_, ok := a.volumes[r.Name]
+	seelog.Infof("Received unmount request %+v", r)
+
+	// Validate the request
+	if len(r.Name) == 0 {
+		return fmt.Errorf("no volume in the request")
+	}
+	if len(r.ID) == 0 {
+		return fmt.Errorf("no mount ID in the request")
+	}
+
+	// Acquire write lock
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	// Find  the volume
```

Comment:
nit: possibly extra whitespace

Reply:
You mean above this comment? I added the newline above it on purpose...

Reply:
Errr let me fix the double space

Reply:
Fixed